app/models/amazon_agent_manager.rb

```diff
@@ -0,0 +1,393 @@
+      @agent_ids << id if ec2.instance(id).exists?
+    end
+
+    @agent_ids
+  end
+
+  def request_queue_empty?
+    messages_in_queue(request_queue).zero?
+  end
+
+  def reply_queue_empty?
+    messages_in_queue(reply_queue).zero?
+  end
+
+  def messages_in_queue(q_name)
+    q = sqs.get_queue_by_name(queue_name: q_name)
+    q.attributes["ApproximateNumberOfMessages"].to_i + q.attributes["ApproximateNumberOfMessagesNotVisible"].to_i
+  rescue => err
+    _log.warn(err.message)
+    0
+  end
+
+  def deploying?
+    !!@deploying
+  end
+
+  def setup_agent
+    agent_ids.empty? ? deploy_agent : activate_agent
+  rescue => err
+    _log.error("No agent is set up to prcoess requests: #{err.message}")
+    _log.error(err.backtrace.join("\n"))
+  end
+
+  def activate_agent
+    agent_ids.each do |id|
+      agent = ec2.instance(id)
+      if agent.state.name == "stopped"
+        agent.start
+        agent.wait_until_running
+        _log.info("Agent #{id} is activated to serve requests.")
+        return id
+      else
+        _log.warn("Agent #{id} is in abnormal state: #{agent.state.name}.")
+        next
+      end
+    end
+
+    _log.error("Failed to activate agents: #{agent_ids}.")
+  end
+
+  # check timestamp of heartbeat of agent_id, return true if the last beat time in
+  # in the time interval
+  def agent_alive?(agent_id, interval = 180)
+    bucket = s3.bucket(ssa_bucket)
+    return false unless bucket.exists?
+
+    obj_id = heartbeat_prefix + agent_id
+    obj = bucket.object(obj_id)
+    return false unless obj.exists?
+
+    last_beat_stamp = obj.last_modified
+    _log.debug("#{obj.key}: Last heartbeat time stamp: #{last_beat_stamp}")
+
+    Time.now.utc - last_beat_stamp < interval && ec2.instance(agent_id).state.name == "running" ? true : false
+  rescue => err
+    _log.error("#{agent_id}: #{err.message}")
+    false
+  end
+
+  def deploy_agent
+    _log.info("Deploying agent ...")
+    @deploying = true
+
+    kp = get_key_pair
+    security_group_id = create_security_group
+    create_setting_yaml
+    data = create_userdata
+    zone_name = ec2.client.describe_availability_zones.availability_zones[0].zone_name
+    subnets = get_subnets(zone_name)
+    raise "No subnet_id is available for #{zone_name}!" if subnets.length.zero?
+    create_profile
+
+    instance = ec2.create_instances({
+      image_id: get_agent_image_id,
+      min_count: 1,
+      max_count: 1,
+      key_name: kp.name,
+      security_group_ids: [security_group_id],
+      user_data: data,
+      instance_type: 't2.micro',
+      placement: {
+        availability_zone: zone_name
+      },
+      subnet_id: subnets[0].subnet_id,
+      iam_instance_profile: {
+        name: profile_name
+      },
+      tag_specifications: [{
+        resource_type: "instance",
+        tags: [{
+          key: "Name",
+          value: MIQ_SSA
+        }]
+      }]
+    })
+    ec2.client.wait_until(:instance_status_ok, {instance_ids: [instance[0].id]})
```

Comment:
Is this a blocking call?  If so, we need to make sure that when this is run in a worker, the worker thread will still get a chance to heartbeat (or perhaps *not* heartbeating is intentional).

Reply:
Yes, it's a blocking call. This is the reason I intentionally set the its worker's heartbeat timeout value to 30.minutes.